cmd/promoted-image-governor/main.go

```diff
@@ -303,6 +309,62 @@ func isMirroredFromOCP(tag api.ImageStreamTagReference, refs []ImageStreamRef) b
 	return false
 }
 
+func deleteTagsOnBuildFarm(ctx context.Context, appCIClient ctrlruntimeclient.Client, buildClusterClients map[string]ctrlruntimeclient.Client, imageStreamsWithPromotedTags map[ctrlruntimeclient.ObjectKey]interface{}, dryRun bool) error {
+	var errs []error
+	for streamKey := range imageStreamsWithPromotedTags {
+		for cluster, client := range buildClusterClients {
+			imageStream := &imagev1.ImageStream{}
+			if err := client.Get(ctx, streamKey, imageStream); err != nil {
+				if !kerrors.IsNotFound(err) {
+					errs = append(errs, fmt.Errorf("could not get image stream %s in namespace %s on cluster %s: %w", streamKey.Name, streamKey.Namespace, cluster, err))
+				} else {
+					logrus.WithField("cluster", cluster).WithField("streamKey", streamKey).Debug("image stream not found")
+				}
+				continue
+			}
+			for _, tag := range imageStream.Status.Tags {
+				imageStreamTag := &imagev1.ImageStreamTag{}
+				tagKey := ctrlruntimeclient.ObjectKey{Namespace: imageStream.Namespace, Name: fmt.Sprintf("%s:%s", imageStream.Name, tag.Tag)}
+				err := appCIClient.Get(ctx, tagKey, imageStreamTag)
+				if err == nil {
+					// the isTag is synced from app.ci
+					continue
+				}
+				if !kerrors.IsNotFound(err) {
+					errs = append(errs, fmt.Errorf("could not get image stream tag %s in namespace %s on cluster %s", tagKey.Name, tagKey.Namespace, appCIContextName))
+					continue
+				} else {
+					// the isTag should be deleted on the build farm since it is not on app.ci
+					logrus.WithField("cluster", appCIContextName).WithField("tagKey", tagKey).Debug("image stream tag not found")
+				}
+				logrus.WithField("cluster", cluster).WithField("tagKey", tagKey.String()).Info("deleting image stream tag on build farm")
+				if dryRun {
+					continue
+				}
+				isTagOnBuildFarm := &imagev1.ImageStreamTag{}
+				if err := client.Get(ctx, tagKey, isTagOnBuildFarm); err != nil {
```

Comment:
Why do you need to GET to DELETE? Could we just DELETE and ignore NotFound?

Reply:
Thought about this too during implementation.
Attempt to avoid delete actions if i can.
I will change accordingly.